decisions/2024-02-28_file_sync.md

```diff
@@ -0,0 +1,91 @@
+
+We can also put in additional logic such as suspending file sync while the main sync is in progress, rate limiting, etc.
+
+It is assumed that we won't do anything special to compress files in MVP.
+A similar process can be implemented for uploading files using (TUS)[tus.io]
+
+![File Sync](./media/file_sync.png)
+
+_Pros:_
+
+- Keeps the main sync process simple/unchanged
+- Allows for targeted logic for file syncing (e.g rate limiting in future)
+- Allows for files to be downloaded where they're needed rather than needing to sync everywhere (via std sync process)
+
+_Cons:_
+
+- We could endup writing quite a bit of code to implement this and may have unexexpcted challenges with the implementation
+- Other sync protocols might provide benefits e.g. rsync for example uses a delta transfer algorithm which only sends the differences between files, rather than the whole file.
+
+
+### Option 2 - Base64 encode files and use std sync
+
+Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simiplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.
+
+This option is probably only viable if we built in some kind of sync record prioritisation, which would make the process more complex.
+
+_Pros:_
+
+- Simple to implement (if we ignore the sync record prioritisation bit)
+
+_Cons:_
+
+- Sync would be blocked until all files are fully downloaded
+- Base64 encoding would increase the size of the file by 33% (or more)
+
+### Option 3 - Use a third party file sync process
+
+A tool such as (https://syncthing.net/)[https://syncthing.net/] or even dropbox, could be setup on every installation to sync files. This would be a simple process, but would mean that we'd need to setup and maintain a third party tool on every installation. This would save us development effort on creating our own file sync process.
+
+_Pros:_
+
+- Not much to develop, but there's still complexities around what files are available, how do we communicate this to the client?
+- Syncthing is a peer to peer protocol, so concievably we could sync files between installations without needing to go via the central server. (unlikely in practice, but possible)
+
+_Cons:_
+
+- How would this work with Android? Can dropbox even be accessed from our code base?
+- We'd need to setup and maintain a third party tool on every installation, even if the time taken for this is acceptable, the risk of miss configurations, missed steps in configuration, etc is high.
+- We probably wouldn't have a lot of control over which files go where, which might cause problems with disk usage, espeically on smaller devices.
+
+
+### Option 4 - Only store files on central server and query central directly for files when needed
+
+Would be the simpliest solution, but wouldn't meet our offline first goals
```

Comment:
Would also still need some like file_list to associate things with specific record of specific table? Could embed in file name/path structure perhaps but potentially doesn't scale well.

Basically saying still want `file_list` regardless, makes it abundantly clear when viewing database schema what's happening! This comment probably applies to option 3 and 5 right?

Reply:
For #5 I think yes, in theory with #3 we could just include a file ID in the existing records that are sync'd I guess.
If the file's not there the user would just get an error.